gateway/app/schemas/qodProvisioning.py

```diff
@@ -0,0 +1,226 @@
+        examples=["123456789@domain.com"],
+    )
+
+
+class PhoneNumber(BaseModel):
+    __root__: str = Field(
+        pattern=r"^\+[1-9][0-9]{4,14}$",
+        description="A public identifier addressing a telephone subscription. In mobile networks it corresponds to the MSISDN (Mobile Station International Subscriber Directory Number). In order to be globally unique it has to be formatted in international format, according to E.164 standard, prefixed with '+'.",
+        examples=["+123456789"],
+    )
+
+
+class SingleIpv4Addr(BaseModel):
+    __root__: IPv4Address = Field(
+        description="A single IPv4 address with no subnet mask",
+        examples=["203.0.113.0"],
+    )
+
+
+class DeviceIpv6Address(BaseModel):
+    __root__: IPv6Address = Field(
+        description="The device should be identified by the observed IPv6 address, or by any single IPv6 address from within the subnet allocated to the device (e.g. adding ::0 to the /64 prefix).\n",
+        examples=["2001:db8:85a3:8d3:1319:8a2e:370:7344"],
+    )
+
+
+class Status(Enum):
+    REQUESTED = "REQUESTED"
+    AVAILABLE = "AVAILABLE"
+    UNAVAILABLE = "UNAVAILABLE"
+
+
+class StatusChanged(Enum):
+    AVAILABLE = "AVAILABLE"
+    UNAVAILABLE = "UNAVAILABLE"
+
+
+class Data(BaseModel):
+    provisioningId: ProvisioningId
+    status: Optional[StatusChanged] = None
+    statusInfo: Optional[StatusInfo] = None
+
+
+class EventStatusChanged(CloudEvent):
+    eventData: Data = Field(description="Event details depending on the event type")
+
+
+class DeviceIpv4Addr1(BaseModel):
+    publicAddress: SingleIpv4Addr
+    privateAddress: SingleIpv4Addr
+    publicPort: Optional[Port] = None
+
+
+class DeviceIpv4Addr2(BaseModel):
+    publicAddress: SingleIpv4Addr
+    privateAddress: Optional[SingleIpv4Addr] = None
+    publicPort: Port
+
+
+class DeviceIpv4Addr(BaseModel):
+    __root__: Union[DeviceIpv4Addr1, DeviceIpv4Addr2] = Field(
+        description="The device should be identified by either the public (observed) IP address and port as seen by the application server, or the private (local) and any public (observed) IP addresses in use by the device (this information can be obtained by various means, for example from some DNS servers).\n\nIf the allocated and observed IP addresses are the same (i.e. NAT is not in use) then  the same address should be specified for both publicAddress and privateAddress.\n\nIf NAT64 is in use, the device should be identified by its publicAddress and publicPort, or separately by its allocated IPv6 address (field ipv6Address of the Device object)\n\nIn all cases, publicAddress must be specified, along with at least one of either privateAddress or publicPort, dependent upon which is known. In general, mobile devices cannot be identified by their public IPv4 address alone.\n",
+        examples=[{"publicAddress": "203.0.113.0", "publicPort": 59765}],
+    )
```

Comment:
No need for all this complexity, use a single class with a validator to ensure either `privateAdress` is set or the other two are set.